cpp/ql/src/semmle/code/cpp/security/CommandExecution.qll

```diff
@@ -2,21 +2,44 @@
 
 import cpp
 import semmle.code.cpp.security.FunctionWithWrappers
+import semmle.code.cpp.models.interfaces.SideEffect
 
 /**
  * A function for running a command using a command interpreter.
  */
-class SystemFunction extends FunctionWithWrappers {
+class SystemFunction extends FunctionWithWrappers, ArrayFunction, AliasFunction, SideEffectFunction {
   SystemFunction() {
-    hasGlobalOrStdName("system") or
-    hasGlobalName("popen") or
+    hasGlobalOrStdName("system") or // system(command)
+    hasGlobalName("popen") or // popen(command, mode)
     // Windows variants
-    hasGlobalName("_popen") or
-    hasGlobalName("_wpopen") or
-    hasGlobalName("_wsystem")
+    hasGlobalName("_popen") or // _popen(command, mode)
+    hasGlobalName("_wpopen") or // _wpopen(command, mode)
+    hasGlobalName("_wsystem") // _wsystem(command)
   }
 
   override predicate interestingArg(int arg) { arg = 0 }
+
+  override predicate hasArrayWithNullTerminator(int bufParam) { bufParam = 0 or bufParam = 1 }
```

Comment:
It's of course a matter of preference, but this disjunction can also be written as `bufParam = [0, 1]` (since about a month ago). Apart from saving bytes, it might make the meaning of the code jump of the screen faster because the reader doesn't have to read the words to see that `bufParam` is used twice and that it's an `or` rather than an `and`.

Also below.

Reply:
Cool.  I suspect this will be most valuable when there are more than two cases (`MagicConstants.qll`!), and I shall look out for opportunities to use it in future.  We _could_ also update the entire models library, but I'm not in a rush to.